pep-9999.rst

```diff
@@ -0,0 +1,217 @@
+
+Examples of such configuration files:
+
+* Pip's ``pip.conf`` and ``pip.ini``
+* Packaging's ``.pypirc`` (also used by tools like Flit and Twine)
+* Poetry's ``config.toml`` and ``auth.toml``
+
+
+Rationale
+=========
+
+While this PEP aim to come up with a unified PyPI repository configuration, thought has been given so to be able to extend the configuration later. For example, PyPI repositories are assumed in this PEP, but a ``type`` attribute could be introduced which would default to ``pypi`` down the line, to allow for other package repository types. Any tool could also start taking advantage of the fact that the configuration files and their discovery is a standard, and use this for their own configuration.
+
+
+Specification
+=============
+
+The specification consists of three parts: the configuration definition, package discovery and configuration file discovery.
+
+
+Configuration definition
+------------------------
+
+The configuration is split into two files:
+
+* ``config.toml`` for PyPI repository configuration
+* ``auth.toml`` for credentials
+
+
+The config.toml file
+''''''''''''''''''''
+
+Package index configuration is defined in the ``config.toml`` file, under the ``[repositories]`` section::
+
+
+    [repositories]
+
+    [repostiories.pypi]
+    url = "https://pypi.org/simple/"
+
+    [repostiories.myrepo1]
+    url = "http://myserver2.org/simple"
+
+
+A repository has the following attributes::
+
+    ==================  ===========  =============================================
+    Attribute           Type         Description
+    ==================  ===========  =============================================
+    final               bool         Make it impossible to override/remove the repository configuration (default: false)
+    proxy_port          int          The port of the proxy server to use (default: 80)
+    proxy_url           str          The URL to the proxy server
+    shadow              bool         Use package from this index, even if it was found in a previous index (default: false)
+    url                 str          URL to repository
+    ==================  ===========  =============================================
+
+
+The auth.toml file
+''''''''''''''''''
+
+Credentials are stored in the ``auth.toml`` file, under the ``[authentication]`` section. There is a correlation between the repository name and the authentication section's name::
+
+    [authentication]
+
+    [authentication.pypi]
+    username = "__token__"
+    password = "pypi-AgEIcHlwaS5vcmcCJDJjMmI5NDJmLWIzNzQtNDc3ZC04ZmRkLTM4NDEzY2M1ZTU5MAACJXsicGVybWlzc2lvbnMiOiAidXNlciIsICJ2ZXJzaW9uIjogMX0AAAYgV91r8iEUUz7Q5xbsfi_KGiO3wrJR2_IOOpnFyLo62bk"
+
+    [authentication.myrepo1]
+    username = "foo"
+    password = "YmFyCg=="
+
+
+Instead of specifying the credentials in this file, Keyring can be used::
```

Comment:
This seems like a requirement applying to tools, rather than part of the specification of the file.

Maybe phrase it as "Tools preferring to look up credentials from other locations, such as `keyring`, are advised to use the full repository URL as the key, rather than the index name or hostname."?

Reply:
Hmm. I need to look into keyring a bit more, I guess. I haven't used it.

So, how would one configure a repository which requires keyring then?
Would they not enter a username and password at all, you mean?

Reply:
In general, tools that support reading from keyring always need a fallback to prompt the user. In that case, they'll write the password back into keyring (if they want, and there's no reason to force that upon them).

So we specify that they should write it back using the full URL, so that provided the user keeps using the full URL all the tools will be looking/setting the same key. That's the best we can do here.

Reply:
Ok, I used your suggestion for now, worded as-is. 👍